xds/internal/balancer/cdsbalancer/cdsbalancer.go

```diff
@@ -278,192 +264,247 @@ func buildProviderFunc(configs map[string]*certprovider.BuildableConfig, instanc
-		// instance is not found in the bootstrap config, we need to put the
-		// channel in transient failure.
-		b.logger.Warningf("Received Cluster resource contains invalid security config: %v", err)
-		b.handleErrorFromUpdate(err, false)
-		return
+	// The errors checked here should ideally never happen because the
+	// ServiceConfig in this case is prepared by the xdsResolver and is not
+	// something that is received on the wire.
+	lbCfg, ok := state.BalancerConfig.(*lbConfig)
+	if !ok {
+		b.logger.Warningf("Received unexpected balancer config type: %T", state.BalancerConfig)
+		return balancer.ErrBadResolverState
+	}
+	if lbCfg.ClusterName == "" {
+		b.logger.Warningf("Received balancer config with no cluster name")
+		return balancer.ErrBadResolverState
 	}
 
-	// The first good update from the watch API leads to the instantiation of an
-	// cluster_resolver balancer. Further updates/errors are propagated to the existing
-	// cluster_resolver balancer.
-	if b.childLB == nil {
-		childLB, err := newChildBalancer(b.ccw, b.bOpts)
-		if err != nil {
-			b.logger.Errorf("Failed to create child policy of type %s: %v", clusterresolver.Name, err)
-			return
-		}
-		b.childLB = childLB
-		b.logger.Infof("Created child policy %p of type %s", b.childLB, clusterresolver.Name)
-	}
-
-	dms := make([]clusterresolver.DiscoveryMechanism, len(update.updates))
-	for i, cu := range update.updates {
-		switch cu.ClusterType {
-		case xdsresource.ClusterTypeEDS:
-			dms[i] = clusterresolver.DiscoveryMechanism{
-				Type:                  clusterresolver.DiscoveryMechanismTypeEDS,
-				Cluster:               cu.ClusterName,
-				EDSServiceName:        cu.EDSServiceName,
-				MaxConcurrentRequests: cu.MaxRequests,
-			}
-			if cu.LRSServerConfig == xdsresource.ClusterLRSServerSelf {
-				bootstrapConfig := b.xdsClient.BootstrapConfig()
-				parsedName := xdsresource.ParseName(cu.ClusterName)
-				if parsedName.Scheme == xdsresource.FederationScheme {
-					// Is a federation resource name, find the corresponding
-					// authority server config.
-					if cfg, ok := bootstrapConfig.Authorities[parsedName.Authority]; ok {
-						dms[i].LoadReportingServer = cfg.XDSServer
-					}
-				} else {
-					// Not a federation resource name, use the default
-					// authority.
-					dms[i].LoadReportingServer = bootstrapConfig.XDSServer
-				}
-			}
-		case xdsresource.ClusterTypeLogicalDNS:
-			dms[i] = clusterresolver.DiscoveryMechanism{
-				Type:        clusterresolver.DiscoveryMechanismTypeLogicalDNS,
-				Cluster:     cu.ClusterName,
-				DNSHostname: cu.DNSHostName,
-			}
-		default:
-			b.logger.Infof("Unexpected cluster type %v when handling update from cluster handler", cu.ClusterType)
+	// Do nothing and return early if configuration has not changed.
+	if b.lbCfg != nil && b.lbCfg.ClusterName == lbCfg.ClusterName {
+		return nil
+	}
+	b.lbCfg = lbCfg
+
+	// Handle the update in a blocking fashion.
```

Comment:
"In some of the other LB policies that use a serializer and where a non-nil error can be potentially returned, we would need to handle this in a blocking way." - ok good to know. Would you happen to know where this is speced/the technical reason why? The reason I brought this up is I had to rewrite Outlier Detection to update Picker synchronously. I don't know any of the other unstated guarantees here though. Ack on the xDS Client explanation, I brought that up because I was thinking if updating synchronously, could potentially trigger a synchronous picker update, but now I know that it's another rule of the API.

Reply:
"The fact that UpdateClientConnState returns an error is an API detail, and whichever LB policy needs to return an error from this method needs to do so synchronously because that is the only way to do it" - ah, this makes sense.

Reply:
"Technically, it could return a nil error synchronously, and later on update the channel with an error picker (if it doesn't like the update pushed to it)" - yeah this sounds much worse haha.